### ISSUE TITLE
Prometheus - 502 Bad Gateway

### ISSUE THREAD

**Issue by solarmon** [sipcapture/homer#422]
I've installed Homer as per:

https://github.com/sipcapture/homer/wiki/Quick-Install#-docker-install

and it is working and receiving (UDP) HEPv3 packets and I can use the Homer web app to view the captures.

I am now exploring what I can do with this docker based setup. I was looking at Grafana and I (think) I understand that I need to add Prometheus as a datasource to Grafana before I can import any Homer based dashboards.

However, I seem to have some issue with Prometheus. Here are the issues I have observed.

docker ps shows that the container status is stuck in 'Restarting'.

The docker logs for the prometheus container has the following error:

level=error ts=2020-10-13T15:19:01.179Z caller=main.go:758 err="error loading config from \"/etc/prometheus/prometheus.yml\": couldn't load configuration (--config.file=\"/etc/prometheus/prometheus.yml\"): open /etc/prometheus/prometheus.yml: permission denied"

Browsing to the Prometheus address on port 9090 returns a '502 Bad Gateway'

When I try to add Prometheus as a datasource in Grafana, I get a 'HTTP Error Bad Gateway' when I use 'http://localhost:9090' as the URL and 'Server (default' as the Access.

When I use 'Browser' as the Access, I get the error 'Unknown error during query transaction. Please check JS console logs.'.

Please advise what I am doing wrong and how to get the Prometheus container working correctly.

Thank you.

**Comment by solarmon** [comment]
OK, after going around in circles and banging my head on a brick wall trying to understand what is going on, I think I have finally cracked it.

It seems the issue is that the prometheus container can't mount to '/etc/prometheus' because it already exists in the 'prom/prometheus:latest' image.

In docker-compose.yml it has the following config for prometheus volumes:

```
 volumes:
      - ./prometheus/:/etc/prometheus/
      - prometheus_data:/prometheus
```

Once I changed it to:

```
volumes:
      - ./prometheus/prometheus.yml:/etc/prometheus/prometheus.yml
      - prometheus_data:/prometheus
```

the prometheus container started up OK and I'm getting a response on port 9090.

The reason why I think '/etc/prometheus' already exists in the image is because when I list '/etc/prometheus' there are already other files in there:

```
/prometheus $ ls /etc/prometheus
console_libraries  consoles           prometheus.yml
/prometheus $
```

There is another file, 'alert.rules', which I assume I will need to mount specifically for too also:

```
- ./prometheus/alert.rules:/etc/prometheus/alert.rules
```

So now that I got the prometheus container starting up correctly, I now need to add it as a datasource in Grafana.

I was still struggling to get the URL to work, in Browser and Server mode, until I realised that I needed to use the internal docker IP of the container. In this case the following URL was valid and tested OK, in Server mode:

http://172.18.0.7:9090/

where 172.18.0.7 is the prometheus docker container IP. I'm not sure whether this is going to be static, or whether I should be using a internal/local DNS address for it, if there is such  a thing that gets created when the docker container starts up?

**Comment by solarmon** [comment]
The issues I observed was for a hom7-prom-all version 7.7.x. build.

I noticed that in the latest hom7-prom-all version 7.8.4, the grafana setup already has a prometheus (and loki) datasource configured, with an URL of:

http://prometheus:9090

and this working and the 'prometheus' hostname resolved correctly to the prometheus container IP. 

However, it seems the docker-compose.yml config still has the non-working volume config for prometheus which still needs to be corrected to fixed the issue with mounting to '/etc/prometheus'.
